### PR TITLE
no break after throwing exception

### PR DESCRIPTION
**Pull request by Sandman83**
Compiler yields a warning about "statement not reachable"

### CODE REVIEW

**Comment by Geod24** [comment]
Which compiler did you try?

**Comment by Sandman83** [comment]
DMD64 D Compiler v2.080.1

**Comment by wilzbach** [comment]
CI fail due to trailing whitespace. Could you please add a unittest on how to reproduce the error to your PR, s.t. this is testable by us and doesn't regress (and ideally connect it with a Bugzilla issue - see the dlang-bot comment for details). Thanks!

**Comment by Sandman83** [comment]
Ok. This was not obvious: I tried to write a unittest, however, I had no luck. It seems to compile well, using dmd, and to yield a warning using dub. 

Given this

```
struct S
{
	@trusted void error(string msg)
	{
		throw new Exception("");
	}

	void fun2(){}

	void fun1()
	{
		error("");
		fun2; 
	}
}
```

`
dmd source/app.d -profile -main
`

works well, however

`
dub build --build=profile
`

yields a warning "statement is not reachable" at the fun2 call in the last line and refuses to compile. 

It affects only the profile build, using together with dub. I tried all by default existing profiles, which dub mentions on its documentation site https://code.dlang.org/docs/commandline


dmd --version
DMD64 D Compiler v2.080.1

dub --version
DUB version 1.9.0, built on May  2 2018


So... I think you can just reject my PR, as it does not concern the real bug. 
But something is going wrong here. I think, it is a feature, that while building via dub it is recognized, that the fun2 call is not reachable (although, only in one of all possible build types). But how can this feature be achieved at all, if even dmd doesn't recognize this circumstance? And if dmd is meant to recognize such circumstances, then, maybe, the example above could be used to find another bug, which has still to be reported?

**Comment by Geod24** [comment]
Added your test case to an existing issue: https://issues.dlang.org/show_bug.cgi?id=13165#c3

File: std/concurrency.d [hunk 1]
```diff
@@ -1986,8 +1988,15 @@ private
                             return true;
                         }
                     }
+                    else
+                    {
+                        if (i + 1 == Ops.length)
+                        {
+                            return false;
+                        }
```
Comment (wilzbach) [inline std/concurrency.d]: @Sandman83 why did you add this check?
`i` can never be bigger than `Ops.length - 1`...

Comment (Sandman83) [reply std/concurrency.d]: If I remember correctly, compiler complained about, that the function doesn't return a value otherwise

File: std/concurrency.d [hunk 2]
```diff
@@ -1967,9 +1967,11 @@ private
                 enum timedWait = false;
             }
 
+            static assert(Ops.length);
+
             bool onStandardMsg(ref Message msg)
             {
-                foreach (i, t; Ops)
+                static foreach (i, t; Ops)
```
Comment (Basile-z) [inline std/concurrency.d]: is this newly `static` foreach run with more iterations than a one at some point ? 
I ask because the two declarations that follow should create an error in this case.

Comment (Sandman83) [reply std/concurrency.d]: Ah... I see. I didn't have any case so far... But nevertheless, it would be wrong then.

**Comment by Sandman83** [comment]
> There's probably a problem [here](https://github.com/dlang/phobos/pull/6621/files#r250245833) to solve

Should I simply close the commit, then? As you pointed out, the static foreach is out of place. But then, none of the changes in std.concurrency are needed?

**Comment by Basile-z** [comment]
I just wanted to avoid any accidental merge after seeing the declarations in the non scoped static foreach.
I'm not saying that you should close.

**Comment by jblachly** [comment]
What's the story with this fix?

`std.regex`:

https://github.com/dlang/phobos/blob/0e8d1a038b222917b7b8ecad3aa0ed1bda5aebbb/std/regex/internal/parser.d#L845

triggers this warning in the recent DMD2 (`DMD64 D Compiler v2.084.1`)

`/Library/D/dmd/src/phobos/std/regex/internal/parser.d(845,13): Warning: statement is not reachable`

**Comment by ljmf00** [comment]
I guess this should be merge. Probably its waiting for @MartinNowak to review, otherwise, people can't use some regex features on DMD compiler.

**Comment by ljmf00** [comment]
@wilzbach can you resolve this?

**Comment by wilzbach** [comment]
> @wilzbach can you resolve this?

Rebased, but it looks like this the `break` has already been removed in master.

File: std/concurrency.d (hunk 1 above)
Comment (wilzbach) [reply std/concurrency.d]: That's weird and shouldn't be the case. Do you have a test case?

Comment (Sandman83) [reply std/concurrency.d]: This was the case while compiling with -profile, if I remember correctly.
Can't reproduce it right now, however...

**Comment by ljmf00** [comment]
@wilzbach sure, it was on May 15 #7004. Sorry. So, it should be closed?
